### PR TITLE
Override `torch.matmul` to cast second array to torch tensor

### PR DESCRIPTION
**Pull request by AlbertMitjans**
Use single dispatch to avoid the following error:

```pycon
>>> m1 = torch.tensor([[1, 0], [0, 1]])
>>> m2 = np.array([[1, 2], [3, 4]])
>>> qml.math.matmul(m1, m2)
TypeError: matmul(): argument 'other' (position 2) must be Tensor, not numpy.ndarray
```

ALSO removed useless tests for the `qml.math.size` function (only the `torch` version is overriden)

### CODE REVIEW

File: pennylane/math/single_dispatch.py [hunk 1]
```diff
@@ -645,6 +646,19 @@ def _sum_torch(tensor, axis=None, keepdims=False, dtype=None):
 ar.register_function("torch", "cond", _cond)
 
 
+def _matmul_torch(tensor1, tensor2, out=None):
```
Comment (albi3ro) [inline pennylane/math/single_dispatch.py:649]: So I think this is actually the case where we should be using `multidispatch`.  The multi-dispatch dispatch figures out the best interface for a multi-interface situation. So if both numpy and torch are ever used in conjunction, it uses torch to perform operations.

Comment (AlbertMitjans) [reply pennylane/math/single_dispatch.py:649]: I believe this is already done by autoray. If you do:
```pycon
>>> t1 = torch.tensor([1])
>>> t2 = [1]
>>> type(qml.math.matmul(t2, t1))
torch.Tensor
```
and viceversa.

My understanding was that we should use `multi_dispatch` when we want to define logic for multiple interfaces, and `single_dispatch` when defining the logic for a single interface.

Comment (josh146) [reply pennylane/math/single_dispatch.py:649]: @AlbertMitjans oh that must be new 🤔 Autoray used to only ever dispatch on the first argument of all functions

Comment (AlbertMitjans) [reply pennylane/math/single_dispatch.py:649]: @josh146 you're right! In the example above `np.matmul` is called, but `numpy` casts the result to a torch Tensor. I will add it to `multi_dispatch` then.

Comment (AlbertMitjans) [reply pennylane/math/single_dispatch.py]: Done @albi3ro

File: tests/math/test_functions.py [hunk 2]
```diff
@@ -2466,24 +2466,6 @@ class TestSize:
         ([[0], [1], [2], [3], [4], [5]], 6),
     ]
 
-    @pytest.mark.autograd
```
Comment (albi3ro) [inline tests/math/test_functions.py:2469]: Why are these getting deleted? Is this a merge conflict gone wrong?

Comment (AlbertMitjans) [reply tests/math/test_functions.py:2469]: Nope, but the feature was only added for torch. These tests are meaningless

**Review by albi3ro — APPROVED** [review]

**Pull request merged**